### PR TITLE
add `tigerbeetle inspect integrity` subcommand

### PR DESCRIPTION
**Pull request by cb22**
This adds a new `tigerbeetle inspect integrity` subcommand, that allows you to verify offline that a datafile is fully intact and has no corruption.

It checks all zones, as well as the grid, for which it reuses the `GridScrubber`. It doesn't check unused parts of the data file, that is padding, free blocks and the like are all ignored.

Example usage:
```
$ tigerbeetle inspect integrity datafile
2025-10-23 19:52:44.657Z info(io): opening "datafile"...
2025-10-23 19:52:44.660Z info(integrity): superblock opened and checked
2025-10-23 19:52:45.289Z info(integrity): successfully checked 1024 wal headers and prepares
2025-10-23 19:52:45.290Z info(integrity): successfully checked 64 client replies
2025-10-23 19:52:45.290Z info(integrity): checking 10018 grid blocks with seed 1916335822393700544...
<fancy zig progress bar>
2025-10-23 19:52:56.063Z info(integrity): successfully checked 10018 grid blocks in 10773ms. (455MiB/s)
```

### CODE REVIEW

File: src/tigerbeetle/cli.zig [hunk 1]
```diff
@@ -287,6 +287,20 @@ const CLIArgs = union(enum) {
         verbose: bool = false,
     };
 
+    // Experimental: the interface is subject to change.
+    const Scrub = struct {
+        log_debug: bool = false,
+        seed: ?[]const u8 = null,
+        memory_lsm_manifest: ?ByteSize = null,
```
Comment (cb22) [inline src/tigerbeetle/cli.zig]: I'd love to get rid of this parameter and fold it in, like how `storage_size_max` is determined from the actual data file size for the scrubber, but I'm not familiar enough with our node pool currently!

**Comment by matklad** [comment]
We probably want a different name, if I understand the meaning of `scrubbing` word correctly? Scrubber actively repairs the grid, it scrubs the stains away. This command in contrast detects stains, but doesn't really remove them. 

`verify` is an easy neutral name, but the problem with that is that we might want to verify many different things. 

Another option is to avoid introduction of a new command and use something like `tigerbeetle inspect consistency`

**Comment by cb22** [comment]
Yeah, I kinda like `tigerbeetle inspect consistency`!

File: src/integration_tests.zig [hunk 2]
```diff
@@ -267,6 +267,11 @@ test "benchmark/inspect smoke" {
         },
     );
 
+    try shell.exec("{tigerbeetle} scrub {data_file}", .{
```
Comment (cb22) [inline src/integration_tests.zig]: (I'm going to add an integration test here to test the negative too, ie corruption in each zone is detected correctly)

File: src/vsr/grid_scrubber.zig [hunk 3]
```diff
@@ -39,7 +39,7 @@ const BlockPtr = @import("./grid.zig").BlockPtr;
 const ForestTableIteratorType = @import("../lsm/forest_table_iterator.zig").ForestTableIteratorType;
 const TestStorage = @import("../testing/storage.zig").Storage;
 
-pub fn GridScrubberType(comptime Forest: type) type {
+pub fn GridScrubberType(comptime Forest: type, grid_scrubber_reads_max: comptime_int) type {
```
Comment (sentientwaffle) [inline src/vsr/grid_scrubber.zig:42]: I think parameterizing this by by `grid_scrubber_reads_max` means that we would be compiling `GridScrubber` into our binary twice, which I'd like to avoid.

Maybe instead we could define the `IOPSType` with the higher of the two values (`constants.grid_iops_read_max`) but then allow overriding with a lower runtime limit via the options.

Or, (and maybe this would be better as a TODO?) our `IOPSType` should not be parameterized with a comptime capacity, but instead should take capacity is runtime and allocate on the heap. That would allow us to tune iops at runtime via CLI flags. That seems potentially useful since TB runs on lots of different hardware, so optimal iops will vary.

File: src/vsr/grid_scrubber.zig [hunk 4]
```diff
@@ -362,6 +364,27 @@ pub fn GridScrubberType(comptime Forest: type) type {
             scrubber.reads_done.push(read);
         }
 
+        /// Like read_fault, but returns information on all reads not just repairs.
+        pub fn read_result_next(scrubber: *GridScrubber) ?union(enum) {
```
Comment (sentientwaffle) [inline src/vsr/grid_scrubber.zig]: I think to keep the scrubber's interface narrow we should just remove `read_fault` entirely:
- return `?union(enum) { {repair, ok}: BlockId }` from `read_result_next`, and
- let the replica skip past `ok` reads itself.

Though actually returning `?struct { block: BlockId, status: enum { ok, repair } }` might be tidier.

File: src/vsr/grid_scrubber.zig [hunk 5]
```diff
@@ -534,11 +557,16 @@ pub fn GridScrubberType(comptime Forest: type) type {
 
             // Wrap around to the next cycle.
```
Comment (sentientwaffle) [inline src/vsr/grid_scrubber.zig]:
```suggestion
```

File: src/tigerbeetle/scrub.zig [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+pub fn command_scrub(
+    gpa: std.mem.Allocator,
+    io: *IO,
+    tracer: *Tracer,
+    args: *const cli.Command.Scrub,
+) !void {
+    var scrub: Scrub = undefined;
+    var scrubbed_bytes: u64 = 0;
+
+    // The superblock is checked as part of initalizing the scrubber - it cannot be skipped.
```
Comment (sentientwaffle) [inline src/tigerbeetle/scrub.zig]:
```suggestion
    // The superblock is checked as part of initializing the scrubber - it cannot be skipped.
```

File: src/tigerbeetle/scrub.zig [hunk 7]
```diff
@@ -0,0 +1,510 @@
...
+
+const Forest = StateMachine.Forest;
+const CheckpointTrailer = vsr.CheckpointTrailerType(Storage);
+const SuperBlock = Replica.SuperBlock;
+
+/// Special offline GridScrubber type - uses constants.grid_iops_read_max rather than
+/// constants.grid_scrubber_reads_max.
+const GridScrubber = vsr.GridScrubberType(Forest, constants.grid_iops_read_max);
+
+pub fn command_scrub(
```
Comment (sentientwaffle) [inline src/tigerbeetle/scrub.zig]: Right now, this immediately panics on the first corruption it finds (e.g. `assert(wal_header.checksum == wal_prepare.checksum);`).

What if instead, we always scrub the whole file, logging errors as we go, and then when we're done exit with a nonzero code if any errors were found?

(e.g. imagine you are scrubbing a huge data file and hit a corrupt block when you are 99% done. You don't know if there is just the one failure or if the last 1% of the data file is garbage.)

File: src/tigerbeetle/scrub.zig [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+    gpa: std.mem.Allocator,
+    io: *IO,
+    tracer: *Tracer,
+    args: *const cli.Command.Scrub,
+) !void {
+    var scrub: Scrub = undefined;
+    var scrubbed_bytes: u64 = 0;
+
+    // The superblock is checked as part of initalizing the scrubber - it cannot be skipped.
+    const scrubbed_bytes_superblock = try Scrub.init(
```
Comment (sentientwaffle) [inline src/tigerbeetle/scrub.zig]: Returning this from `Scrub.init` seems awkward to me... I think we could just add it as

```
    for (scrub.superblock.reading) |_| {
        scrubbed_bytes += vsr.superblock.superblock_copy_size;
    }
```

here in `command_scrub`. Or alternatively, we don't consider this "scrubbing" (since we don't check all copies anyway) and instead just say that an openable superblock is a prerequisite for scrubbing, and an unopenable superblock is a fatal error.

File: src/tigerbeetle/scrub.zig [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+            scrub.buffer_prepare[0..@sizeOf(vsr.Header)],
+        ));
+
+        const wal_prepare_body_valid =
+            wal_prepare.valid_checksum() and
+            wal_prepare.valid_checksum_body(
+                scrub.buffer_prepare[@sizeOf(vsr.Header)..wal_prepare.size],
+            );
+
+        assert(wal_header.valid_checksum() and wal_prepare_body_valid);
```
Comment (sentientwaffle) [inline src/tigerbeetle/scrub.zig]: I'd split these so that if it hits this assertion, we have more information about the issue. (Also more readable, as a bonus.)

```suggestion
        assert(wal_header.valid_checksum());
        assert(wal_prepare_body_valid);
```

File: src/tigerbeetle/scrub.zig [hunk 10]
```diff
@@ -0,0 +1,510 @@
...
+
+io: *IO,
+storage: Storage,
+
+superblock: SuperBlock,
+client_sessions_checkpoint: CheckpointTrailer,
+grid: Grid,
+forest: Forest,
+grid_scrubber: GridScrubber,
+grid_blocks_scrubbed: std.AutoArrayHashMapUnmanaged(u64, void),
```
Comment (sentientwaffle) [inline src/tigerbeetle/scrub.zig]: It seems like this could be a bitset instead? Though hash map is fine too, whichever you prefer.

File: src/tigerbeetle/scrub.zig (hunk 8 above)
Comment (cb22) [reply src/tigerbeetle/scrub.zig]: Yeah, it is a little awkward, but I really wanted to keep the scrubbed bytes calculation both split (eg, the actual and the target are calculated separately) and local (immediately after checking said blocks, increment the counter).

I think I'm going to make `open()` also assert the superblock is open and verified, and it can return the bytes checked.

File: src/tigerbeetle/scrub.zig (hunk 10 above)
Comment (cb22) [reply src/tigerbeetle/scrub.zig]: Yep! A previous iteration of the code made more sense with a hash map, but I no longer need one.

File: src/tigerbeetle/scrub.zig (hunk 7 above)
Comment (cb22) [reply src/tigerbeetle/scrub.zig]: I was back and forth on this a little. 

On the one hand, dealing with corruption means that percentages can be meaningless in the grid; if you have a corrupt index block, for example, the data blocks might be completely fine but you can't check them. The main goal of the subcommand so far is to answer the question "is this data file perfect". 

However!

Having the functions return errors means that it then becomes possible to unit test them easily, which means we don't need a whole bunch of slow integration tests for different types of corruption, which I think is very worthwhile. I'm not yet sure if I'll make the scrubbing continue after finding the first error, but I'll see once I've reworked them.

File: src/shell.zig [hunk 11]
```diff
@@ -513,11 +513,11 @@ fn exec_inner(
     defer if (poller) |*p| p.deinit();
 
     errdefer |err| {
-        log.err("process failed with {s}: {s}", .{ @errorName(err), argv_formatted });
+        log.info("process failed with {s}: {s}", .{ @errorName(err), argv_formatted });
         if (poller) |*p| {
             inline for (comptime std.enums.values(Streams)) |stream| {
                 if (p.fifo(stream).count > 0) {
-                    log.err("{s}:\n++++\n{s}++++\n", .{
+                    log.info("{s}:\n++++\n{s}++++\n", .{
```
Comment (sentientwaffle) [inline src/shell.zig]: :thinking: Why change this?

**Review by sentientwaffle — COMMENTED** [review]
Integration tests are failing with

```
error: 'integration_tests.test.benchmark/inspect smoke' failed: expected error.ExecFailed, found error.ExecNonZeroExitStatus
D:\a\tigerbeetle\tigerbeetle\src\shell.zig:569:42: 0x7ff710748f36 in exec_inner (test-integration.exe.obj)
        .Exited => |code| if (code != 0) return error.ExecNonZeroExitStatus,
                                         ^
```

File: src/shell.zig (hunk 11 above)
Comment (cb22) [reply src/shell.zig]: > error: 'integration_tests.test.benchmark/inspect smoke' failed: expected error.ExecFailed, found error.ExecNonZeroExitStatus

Yeah... Shell's exec returns a different error on different platforms!

> 🤔 Why change this?

Related - we want the process to exit unsuccessfully, but error logs force a test failure (and warns show up in the output regardless). I'm probably going to use ChildProcess directly, or something similar, and revert this!

File: src/tigerbeetle/scrub.zig (hunk 7 above)
Comment (cb22) [reply src/tigerbeetle/scrub.zig]: In the end, I decided not to refactor it out for now and stick with the original design! For backup verification, it's most important to know that the file is completely intact, but I think we can improve this down the line.

**Review by sentientwaffle — APPROVED** [review]
